### PR TITLE
Implement -Wvalue-discard warning

### PR DESCRIPTION
**Pull request by cb372**
Warn when an expression `e` with non-Unit type is adapted by embedding it into a block `{ e; () }` because the expected type is Unit.

In Scala 2.13 ([relevant code in Typers.scala](https://github.com/scala/scala/blob/d578a02ea6b41b662072759c82c19f9309a15176/src/compiler/scala/tools/nsc/typechecker/Typers.scala#L1116)) there are two special cases:

1. The warning can be disabled by explicitly annotating an expression with `: Unit`.
    * In this PR I've elected not to implement this, as IMO it's not necessary now we have `@nowarn`. It's just as easy to annotate with `@nowarn` as with `: Unit`.
2. No warning is raised if the expression is an application of a function that returns `this.type`, because this is a very common operation when using mutable collections.
    * The heuristic is actually slightly more subtle than that. See the `warn-value-discard.scala` test for some examples.
    * I have implemented this special case by copying the relevant code from the `2.13.x` branch in scala/scala, so the behaviour should be equivalent to Scala 2.13.

### CODE REVIEW

**Comment by som-snytt** [comment]
It's documented somewhere, but Scala 2 avoids warning for methods that return `this.type`.
```
def append(x: X): Unit = buf.addOne(x)
```
The user can "annotate" that discarding is desired:
```
def f(): Unit = x.m(): Unit
```
That convention was devised before `@nowarn` and `-Wconf``, so I don't know what would be best now; the "explicit annotation" exception means a warning is never generated, instead of generating a warning that must be post-processed.

I'd suggest not supporting ancient compiler option aliases (the original `-Ywarn`).

If they want to support `-Xlint` (or `-Wlint`) as an option, it's worth considering calling it `-Xlint:value-discard`, where the distinction between `-W` and `-Xlint` is that `-W` warns about dubious language features or interactions (`-Wnumeric-widen`), whereas `-Xlint` warns about code that may be reasonable. (Maybe `-Wvalue-discard` is the better name under that criterion, who knows?)

**Comment by odersky** [comment]
This is a fine line to walk since lots of parts of the Scala library do return something that can safely be ignored. For instance
`put` or `remove` on mutable collections, or the many functions that return `this`. 

To make the test bullet-proof it would need to be augmented with effect checking. It makes indeed no sense to discard the value of a pure expression, but for side-effecting expressions it's common practice.

**Comment by SethTisue** [comment]
> This is a fine line to walk since lots of parts of the Scala library do return something that can safely be ignored

Note that in Scala 2 these warnings aren't enabled by default, and they aren't enabled by `-Xlint` either. Users must explicitly opt-in, and it's expected that pure-functional coders are the main audience. Those coders are already avoiding "parts of the Scala library" such as mutable collections. (And in places where they must use those parts, they don't mind a little extra ceremony.)

> the many functions that return `this`

I agree with Som that these are important to exempt.

**Comment by som-snytt** [comment]
Relatedly, `-Wnonunit-statement` resulted in narrowing the mutators of `StringBuilder` to `this.type` in https://github.com/scala/scala/pull/10047 which suggests that adding an annoying tool to the tool belt may result in useful behavior changes to avoid the annoyance.

My recent interaction with `ConcurrentMap` and the Java collection wrappers was a reminder that the half-life of a collections API may be half as long as previously estimated.

**Comment by gvolpe** [comment]
Apologies for rekindling this discussion, but I am currently finishing writing a [Scala 3 book](https://leanpub.com/feda) focused on purely functional programming, and I was wondering if there were any plans for Scala 3 to add support for `-Ywarn-value-discard` (the Scala 2 equivalent). This flag is enabled by default in [sbt-tpolecat](https://github.com/typelevel/sbt-tpolecat), a plugin used by most FP applications.

I've been following the work on #16157, but it seems this work is out of scope. Then Chris made me aware of this draft PR. I've also been pointed to the [zerowaste plugin](https://github.com/ghik/zerowaste), but unfortunately it doesn't seem to work. 

I've been told this feature is planned to be implemented after 3.3, but I couldn't find any discussion about it. Is there any link related to this I can refer to in my book?

This flag is a must-have for production systems, and I believe it currently is the biggest blocker to Scala 3 adoption in FP shops.

**Comment by cb372** [comment]
I've ported over from Scala 2 the special-case handling of functions that return `this.type`, so I think this is now ready for a review.

File: compiler/src/dotty/tools/dotc/ast/TreeInfo.scala [hunk 1]
```diff
@@ -308,6 +308,159 @@ trait TreeInfo[T <: Untyped] { self: Trees.Instance[T] =>
...
+   *      * targs = Nil
+   *      * argss = List(List(arg11, arg12...), List(arg21, arg22, ...))
+   *
+   *    Apply(Apply(TypeApply(foo, List(targs1, targs2, ...)), List(arg21, arg22, ...)), List(arg11, arg12...))
+   *      * callee = TypeApply(foo, List(targs1, targs2, ...))
+   *      * core = foo
+   *      * targs = Nil
+   *      * argss = List(List(arg11, arg12...), List(arg21, arg22, ...))
+   */
+  final class Applied(val tree: Tree) {
```
Comment (KacperFKorban) [inline compiler/src/dotty/tools/dotc/ast/TreeInfo.scala:353]: Can we possibly reuse some of the existing utility methods e.g. `allArguments`, `methPart`, `stripApply`, etc.?
Or else make those methods top-level? (Some of them seem unnecessary for this case)

File: compiler/src/dotty/tools/dotc/ast/TreeInfo.scala [hunk 2]
```diff
@@ -308,6 +308,159 @@ trait TreeInfo[T <: Untyped] { self: Trees.Instance[T] =>
...
+        case ThisType(tref) =>
+          tref.symbol == receiver.symbol
+        case tref: TermRef =>
+          tref.symbol == receiver.symbol || argss.exists(_.exists(tref.symbol == _.symbol))
+        case _ =>
+          def checkSingle(sym: Symbol): Boolean =
+            (sym == receiver.symbol) || {
+              receiver match {
+                case Apply(_, _) => op.isOpAssignmentName                     // xs(i) += x
+                case _ => receiver.symbol.isGetter || receiver.symbol.isField // xs.addOne(x) for var xs
```
Comment (KacperFKorban) [inline compiler/src/dotty/tools/dotc/ast/TreeInfo.scala:446]: The equivalent check to scala2's `symbol != null` is `symbol != NoSymbol`.
We should add `receiver.symbol != NoSymbol` here to be safe.

File: compiler/src/dotty/tools/dotc/ast/TreeInfo.scala [hunk 3]
```diff
@@ -308,6 +308,159 @@ trait TreeInfo[T <: Untyped] { self: Trees.Instance[T] =>
...
+            case m: MethodType =>
+              m.resType match {
+                case ThisType(tref) => checkSingle(tref.symbol)
+                case tref: TermRef  => checkSingle(tref.symbol)
+                case restpe => loop(restpe)
+              }
+            case PolyType(_, restpe) => loop(restpe)
+            case _ => false
+          }
+          loop(fun.symbol.info)
```
Comment (KacperFKorban) [inline compiler/src/dotty/tools/dotc/ast/TreeInfo.scala:459]: Same as above `fun.symbol != NoSymbol`

**Review by KacperFKorban — COMMENTED** [review]
I tested the PR and it seems to work as expected (the same as in Scala 2). Added some suggestions/questions about the code.

File: compiler/src/dotty/tools/dotc/ast/TreeInfo.scala (hunk 1 above)
Comment (cb372) [reply compiler/src/dotty/tools/dotc/ast/TreeInfo.scala:371]: Thanks for the pointer. When I was copying swathes of code from Scala 2, I didn't noticed those almost-identical methods already existed. I've refactored `Applied` to make use of them.

File: compiler/src/dotty/tools/dotc/ast/TreeInfo.scala [hunk 4]
```diff
@@ -308,6 +326,132 @@ trait TreeInfo[T <: Untyped] { self: Trees.Instance[T] =>
     case Block(_, expr) => forallResults(expr, p)
     case _ => p(tree)
   }
+
+  /** Applications in Scala can have one of the following shapes:
+   *
+   *    1) naked core: Ident(_) or Select(_, _) or basically anything else
+   *    2) naked core with targs: TypeApply(core, targs) or AppliedTypeTree(core, targs)
```
[the remark this answers is not in the document]
Comment (smarter) [reply compiler/src/dotty/tools/dotc/ast/TreeInfo.scala:333]: (And an AppliedTypeTree can wrap another AppliedTypeTree, e.g. `type Foo[X] = [Y] =>> Y; val x: Foo[Int][String] = ""`)

**Review by KacperFKorban — APPROVED** [review]
@cb372 
Thanks for all your work on this feature ❤️ 

Since we want this feature to be in the next branched-out release, I will merge this PR now and address the remaining review comment in a separate PR immediately following this one (https://github.com/lampepfl/dotty/pull/16637)

**Comment by cb372** [comment]
Thank you! I was planning to address that comment but it was gradually moving further down my todo list 😄 so I'm happy for you to take over from here.

**Pull request merged**